docs/design/jet/025-jet-job-member-pruning.md

```diff
@@ -0,0 +1,192 @@
+
+### Processor pruning
+
+We would like to separate processor pruning into two categories: **intra-member** and **inter-member**.
+
+#### Use cases for processor pruning
+
+##### Processor pruning for Scan -> Transform -> Aggregate
+
+1. Single-staged aggregation.
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tTM.svg)
+
+On the picture above we can see optimized example from member pruning case. We can go further and just don't create 
+processors which are not participating in data processing by tuning local parallelism parameter with 
+partition involvement knowledge: 
+
+![Scan + Scan -> HashJoin](https://svgshare.com/i/tUD.svg)
+
+So, it's a good target for **intra-member** processor pruning.
+
+2. Double-staged aggregation.
+```
+- COMBINE
+  - ACCUMULATE
+    - TRANSFORM
+      - SCAN[map, partitionKey=K]
+```
+![Scan + Scan -> HashJoin](https://jet-start.sh/docs/assets/arch-dag-4.svg)
+
+Two-stage aggregation might be a good target for **inter-member** processor pruning. However,
+processor logic is opaque to Jet and changing it requires a lot of changes, which may jeopardize
+the correctness of execution plans. Potentially, it may be done in another way: we can translate
+two-staged into single-staged on SQL opt phase and then eliminate member(s) via member pruning,
+namely scan and flatmap nodes.
+
+#### Solution design details
+
+##### Intra-member processor pruning
+To control processor creation and parallelism within one member, we would like to use various processor suppliers
+(`ProcessorMetaSupplier` or `ProcessorSupplier`). The correctness of this method will totally rely on how DAG is constructed.
+Since Partition Pruning initiative was introduced to align PredicateAPI and SQL functionality and performance, we will
+rely on SQL optimizer input and DAG construction phase in `CreateDagVisitor`. 
+This approach was tried, but it was **rejected** due to **small performance difference for increased code complexity**.
+##### Inter-member processor pruning
+
+After long discussions, we decided NOT to support this kind of processor pruning, because it
+- is relatively ineffective since most DAGs with broadcast edges were created with algorithm correctness in mind,
+- is hard to implement for most use cases, and
+- doesn't fit the goal of general effort.
+
+### Scan processor partition pruning
+
+Partition pruning is a pretty simple optimization: we will extract partition key condition during SQL opt phase,
+pass it to specialized processor supplier which will spawn `ReadMapOrCacheP` with only required partitions to scan.
+
+## Final decision
+After long discussions, we decided to implement **SQL-oriented** approach for member pruning and Scan processor partition pruning.
+
+Processor pruning was considered as non-universal, complex and  **rejected**. As a side effect from this research, 
+we decided to add  new `lazyForceTotalParallelismOne` PMS builder which does not cache member address to prevent 
```

Comment:
`lazyForceTotalParallelismOne` is not only lazy, but the main characteristic is that is it partition-assignment aware. you completelty skipped description of that while this is quite important

Reply:
lazyForceTotalParallelismOne deserves it's own section in TDD

Reply:
still waiting for that

Reply:
check now

Reply:
added new comment: https://github.com/hazelcast/hazelcast/pull/24596/files#r1263500386